app/src/main/java/ee/ioc/phon/android/speechutils/ContinuousRawAudioRecorder.java

```diff
@@ -0,0 +1,215 @@
+            if (this == mBeginningOfBufferPosition && mSessionStartPointerMillis == Integer.MIN_VALUE)
+                return mNowPosition.getSessionStartPointerMillis();
+
+            // in case that the requested start pointer is bigger than the buffer, return the buffer size
+            if (this != mBeginningOfBufferPosition && mSessionStartPointerMillis > mBeginningOfBufferPosition.getSessionStartPointerMillis())
+                return mBeginningOfBufferPosition.getSessionStartPointerMillis();
+
+            return mSessionStartPointerMillis;
+        }
+
+        static void setRecordingBufferLengthMillis(int recordingBufferLengthMillis) {
+            mBeginningOfBufferPosition.setSessionStartPointerMillis(recordingBufferLengthMillis);
+        }
+
+        public static SessionStartPointer beginningOfBuffer() {
+            return mBeginningOfBufferPosition;
+        }
+
+        public static SessionStartPointer now() {
+            return mNowPosition;
+        }
+
+        public static SessionStartPointer someMillisBack(int millisBackToStartTheSessionFrom) {
+            return new SessionStartPointer(millisBackToStartTheSessionFrom);
+        }
+
+        public static SessionStartPointer someSecondsBack(int secondsBackToStartTheSessionFrom) {
+            return new SessionStartPointer(secondsBackToStartTheSessionFrom * 1000);
+        }
+    }
+
+    public ContinuousRawAudioRecorder(int audioSource, int sampleRate, int recordingBufferLengthMillis) {
+        super(audioSource, sampleRate, recordingBufferLengthMillis, true);
+
+        // this is very important. We introduce the buffer length to the SessionStartPointer object
+        SessionStartPointer.setRecordingBufferLengthMillis(recordingBufferLengthMillis);
+
+        try {
+            int bufferSize = getBufferSize();
+            int framePeriod = bufferSize / (2 * RESOLUTION_IN_BYTES * CHANNELS);
+            createRecorder(audioSource, sampleRate, bufferSize);
+            createBuffer(framePeriod);
+            setState(State.READY);
+        } catch (Exception e) {
+            if (e.getMessage() == null) {
+                handleError("Unknown error occurred while initializing recorder");
+            } else {
+                handleError(e.getMessage());
+            }
+        }
+    }
+
+    public ContinuousRawAudioRecorder(int sampleRate) {
+        this(DEFAULT_AUDIO_SOURCE, sampleRate, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public ContinuousRawAudioRecorder() {
+        this(DEFAULT_AUDIO_SOURCE, DEFAULT_SAMPLE_RATE, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public String getWsArgs() {
+        return "?content-type=audio/x-raw,+layout=(string)interleaved,+rate=(int)" + getSampleRate() + ",+format=(string)S16LE,+channels=(int)1";
+    }
+
+    public ContinuousRawAudioRecorder setSessionStartPointer(SessionStartPointer sessionStartPointer) {
+        mSessionStartPointer = sessionStartPointer;
+        return this;
+    }
+
```

Comment:
whenever the recording should be consumed, I calculate the number of samples that should be fetched from the latest recording sample (==now). Here, the concept of consumed vs recording session comes to act. The current behaviour is that if the consumed session and the recording session are different (which occurs when the recording pointer passes the consumed pointer), we consume the buffer from what was marked as the session start point (e.g. beginning of the buffer/now/now-X).
In case that the sessions are the same (e.g. we've read everything from the session start point and now it's the 2nd call to consumRecording), I get the data from the consumed pointer (==startPos) to now